draft-ietf-dnsop-svcb-httpssvc.md

```diff
@@ -415,8 +414,13 @@ each of which contains:
   (but constrained by the RDATA and DNS message sizes).
 * an octet string of the length defined by the previous field.
 
-If the parser reaches the end of the RDATA while parsing a SvcFieldValue,
-the RR is invalid and MUST be discarded.
+SvcParamKeys SHALL appear in increasing numeric order.
+
+Clients MUST reject the RR as malformed if
```

Comment:
Should we mention that clients SHOULD instead use other non-malformed RRs in the RRSET if possible?  (We want to avoid forward compat issues where a client implementation bug or some change causes entire rrsets to get rejected.)

Would it be safer for some of these to have deterministic behavior rather than "reject the RR"?

Reply:
If a RR is malformed the entire RRset should be rejected.  One cannot safely use partial RRsets.   Operators should not depend on RRsets getting through a recursive server if a RR is malformed.

STD 13 says to reject the entire zone if a single RR is malformed when loading a zone. 
Recursive servers that are aware of the structure of these records should be rejecting answers that contain malformed RRs (they are not permitted to filter out individual RRs).

Reply:
That would seem to argue in-favor of having deterministic handling of certain types of errors rather than rejecting the RR? 
For example, in the case of a parameter appearing multiple times, clients MUST use only the first instance.
Overly strict rejection (rather than deterministic handling) could make extensibility much harder in the future?

Reply:
There are exactly 4 conditions listed here.

1 the last TLV being malformed (incomplete tag, length or value)
2 TLVs not being sorted on T
3 multiple T with the same value
4 L being incorrect for a given T

Named already implements 1.
Named already implements 4 for some T.

e.g. for port L == 2, for hint4 L != 0 && L % 4 == 0.

This checks are done as named parses a record learn off the wire or after loading with unknown record format.   All the tools which lookup records also perform these checks. 

None of these currently implemented checks actually prevent new Ts being deployed.

Reply:
I would have been worried about invalid SVCBs breaking everything until I came to accept the suggestion from a couple weeks ago that the client would handle an SVCB timeout by disallowing fallback to A/AAAA.  So we're already into the territory that a resolver doing something bad or a bad config can completely break resolving the name, and not just the SVCB response.

Specific to the case of repeating keys, as long as we have good standardized semantics for list values, I don't see any reason we'd ever need to switch to allowing repeated keys.  So I don't see any reason to ever treat it as anything other than malformed and thus reject the RRSet.  Better for completely invalid things to get noticed by brokenness, so people fix them.

But in general, anywhere that we have good reason to want future extensibility, yes, deterministic handling of things that are "valid but unknown" is often reasonable.  The obvious example is the "clients MUST ignore SvcFieldValue parameters that they do not recognize" rule.

Reply:
> If a RR is malformed the entire RRset should be rejected. One cannot safely use partial RRsets.

I think this is reasonable.  I've updated the PR with that logic.

Technically, I don't think it's a security issue, but this situation should never happen so it makes sense to be cautious.  Who knows what else might have gotten corrupted if one RR is malformed?  Rejecting the whole RRSet seems like a clearer signal to the zone owner that something is wrong with (e.g.) their authoritative server.

> Operators should not depend on RRsets getting through a recursive server if a RR is malformed.
> Recursive servers that are aware of the structure of these records should be rejecting answers that contain malformed RRs (they are not permitted to filter out individual RRs).

I disagree with this.  I've updated the PR to clarify that we don't want recursive resolvers even thinking about the TLVs.  If recursive resolvers are dropping whole RRSets, it will be harder for clients to debug this class of problems (SERVFAIL can mean almost anything).  It's also bad for consistency between resolvers, bad for performance, and could encourage laziness on the part of clients.

> That would seem to argue in-favor of having deterministic handling of certain types of errors rather than rejecting the RR?

I agree that determinism is important, and the text was unclear on what was supposed to happen.  I've updated the PR to clarify the intended client behavior.

> For example, in the case of a parameter appearing multiple times, clients MUST use only the first instance.
> Overly strict rejection (rather than deterministic handling) could make extensibility much harder in the future?

I'm open to this kind of change, but I think if we impose the single-value requirement now, the chances of us wanting to remove it later are very slim.  If we did do something like what you describe, I think we would also need to GREASE it, because clients could easily implement the "first value wins" logic incorrectly (e.g. by rejecting the RR, or using the last value instead of the first) without visibly different behavior in any normal case.

As @marka63 notes, extension by adding new types should continue to work as expected, and seems to cover all of our anticipated needs.

Reply:
yeah, it's really the support for new SvcParamTypes that is most important.  (Do we want to encourage greasing of them and allocating a range for greased ones?)

Reply:
> > If a RR is malformed the entire RRset should be rejected. One cannot safely use partial RRsets.
> 
> I think this is reasonable. I've updated the PR with that logic.
> 
> Technically, I don't think it's a security issue, but this situation should never happen so it makes sense to be cautious. Who knows what else might have gotten corrupted if one RR is malformed? Rejecting the whole RRSet seems like a clearer signal to the zone owner that something is wrong with (e.g.) their authoritative server.
> 
> > Operators should not depend on RRsets getting through a recursive server if a RR is malformed.
> > Recursive servers that are aware of the structure of these records should be rejecting answers that contain malformed RRs (they are not permitted to filter out individual RRs).
> 
> I disagree with this. I've updated the PR to clarify that we don't want recursive resolvers even thinking about the TLVs. If recursive resolvers are dropping whole RRSets, it will be harder for clients to debug this class of problems (SERVFAIL can mean almost anything). It's also bad for consistency between resolvers, bad for performance, and could encourage laziness on the part of clients.

Yes, SERVFAIL can mean almost anything which is why there are tools like DiG that are designed to follow delegations and talk directly to authoritative servers.  They can even dump malformed responses as hex blobs.

Named has been rejecting malformed RR content for over a decade now and I guarantee you that it has been a net plus.  I think there has been 1 bug report in that entire time where someone put garbage into a record in a zone and expected named to serve it rather than rejecting it or expected it to get through a resolver. 

I also think that you are worrying way too much about malformed records being there in the first place.  People will use tools to generate the records.  The tools will get it right from the start or will be very quickly fixed to get it right.  SERVFAIL from the recursive server is a great motivator to fix tools.

As for lazy clients, whether the resolver rejects malformed records or not it won't make a difference.

I also don't want to have to do the validation of the record as I am printing it out.  Validation on data entry is much safer.

Reply:
Also if it isn't the recursive server checking the content it will end up being the firewall and once it is validation the record content debugging becomes just about impossible.

Reply:
@marka63 Thanks for sharing your experience with malformed records in practice.  I've removed the recommendation regarding recursive server behavior.

I think that this draft is negotiating a culture clash between the flexibility of DNSOP and the paranoia of TLS, but I'm optimistic that we'll be able to find a middle ground on the question of specificity.